homeassistant/components/apcupsd/__init__.py

```diff
@@ -1,87 +1,156 @@
+VALUE_ONLINE: Final = 8
+PLATFORMS: Final = (Platform.SENSOR, Platform.BINARY_SENSOR)
+MIN_TIME_BETWEEN_UPDATES: Final = timedelta(seconds=60)
 
-VALUE_ONLINE = 8
 
 CONFIG_SCHEMA = vol.Schema(
     {
         DOMAIN: vol.Schema(
             {
-                vol.Optional(CONF_HOST, default=DEFAULT_HOST): cv.string,
-                vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+                vol.Optional(CONF_HOST, default="localhost"): cv.string,
+                vol.Optional(CONF_PORT, default=3551): cv.port,
             }
         )
     },
     extra=vol.ALLOW_EXTRA,
 )
 
 
-def setup(hass: HomeAssistant, config: ConfigType) -> bool:
-    """Use config values to set up a function enabling status retrieval."""
-    conf = config[DOMAIN]
-    host = conf[CONF_HOST]
-    port = conf[CONF_PORT]
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up integration from legacy YAML configurations."""
+    conf = config.get(DOMAIN)
+    if conf is None:
+        return True
+
+    # We only import configs from YAML if it hasn't been imported. If there is a config
+    # entry marked with SOURCE_IMPORT, it means the YAML config has been imported.
+    for entry in hass.config_entries.async_entries(DOMAIN):
+        if entry.source == SOURCE_IMPORT:
+            return True
+
+    # Since the YAML configuration for apcupsd consists of two parts:
+    # apcupsd:
+    #   host: xxx
+    #   port: xxx
+    # sensor:
+    #   - platform: apcupsd
+    #     resource:
+    #       - resource_1
+    #       - resource_2
+    #       - ...
+    # Here at the integration set up we do not have the entire information to be
+    # imported to config flow yet. So we temporarily store the configuration to
+    # hass.data[DOMAIN] under a special entry_id SOURCE_IMPORT (which shouldn't
+    # conflict with other entry ids). Later when the sensor platform setup is
+    # called we gather the resources information and from there we start the
+    # actual config entry imports.
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][SOURCE_IMPORT] = conf
+    return True
 
-    apcups_data = APCUPSdData(host, port)
-    hass.data[DOMAIN] = apcups_data
 
-    # It doesn't really matter why we're not able to get the status, just that
-    # we can't.
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Use config values to set up a function enabling status retrieval."""
+    data_service = APCUPSdData(
+        config_entry.data[CONF_HOST], config_entry.data[CONF_PORT]
+    )
+
     try:
-        apcups_data.update(no_throttle=True)
-    except Exception:  # pylint: disable=broad-except
-        _LOGGER.exception("Failure while testing APCUPSd status retrieval")
+        await hass.async_add_executor_job(data_service.update)
+    # It doesn't really matter why we're not able to get the status, just that we can't.
+    except Exception as ex:  # pylint: disable=broad-except
```

Comment:
We only catch OSError in the config flow.
```suggestion
    except OSError as ex:
```